### ISSUE TITLE
SSL Passthrough

### ISSUE THREAD

**Issue by rusenask** [heptio/contour#15]
Hello, it would be good to know whether Contour supports ssl passthrough and if it doesn't - whether it would be possible/reasonable to add it. 

NGINX ingress controller has this ` ingress.kubernetes.io/ssl-passthrough: "true"` annotation, it might make sense to keep the format the same to make it easier to switch between ingresses.

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    kubernetes.io/ingress.class: contour
    ingress.kubernetes.io/ssl-passthrough: "true"
  name: relay-ingress
  namespace: default
```

**Comment by rothgar** [comment]
related to #14

**Comment by davecheney** [comment]
@rusenask  Thanks for raising this issue. I need to ask a few clarifying questions to check I understand what a successful outcome of this issue would look like.

q. In this scenario who is doing TLS processing? Is it the load balancer in front of Contour? Is it Contour? Is it the service the ingress points to? Is it two or maybe all three?

q. If TLS processing is being done by the final k8s service, it feels like you'd want to configure Envoy as a TCP proxy, not a HTTPS proxy. Did I get that right?

**Comment by rusenask** [comment]
@rothgar that issue should probably be configured through another annotation:
`ingress.kubernetes.io/secure-backend: "false/true"`

@davecheney 
> q. In this scenario who is doing TLS processing? Is it the load balancer in front of Contour? Is it Contour? Is it the service the ingress points to? Is it two or maybe all three?

Service that ingress points to. LB and Contour would just pass through traffic to the destination, no termination should happen. 

> q. If TLS processing is being done by the final k8s service, it feels like you'd want to configure Envoy as a TCP proxy, not a HTTPS proxy. Did I get that right?

Not sure about Envoy TCP proxy but what I disliked in nginx-ingress was that if you set it to TCP proxy then it can't match hostnames anymore, it's more like 8080:namespace/serviceName:port:

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: tcp-configmap
  namespace: kube-system
data:
  8080: "default/whr:8080"  
```

Desired behaviour in this case would be that you could still route based on hostnames to different services and ports. 

I hope this makes sense.

**Comment by davecheney** [comment]
> Service that ingress points to. LB and Contour would just pass through traffic to the destination, no termination should happen.

Ingress objects are strongly coupled to HTTP. This sounds like a good use case for using a service load balancer directly. Have you tried that? What are the reasons why you decided not to use this approach?

> Not sure about Envoy TCP proxy but what I disliked in nginx-ingress was that if you set it to TCP proxy then it can't match hostnames anymore, it's more like 8080:namespace/serviceName:port:

If you're in TCP mode, then i'm guessing there is no HTTP request to inspect to route the traffic. It sounds like you have to dedicate the public IP to forwarding the traffic to only one service, which is very similar to the service load balancer described above.

While I think Envoy can do this, I have some doubts about how to express this in the limitations of the Ingress object.

**Comment by davecheney** [comment]
@rusenask ping, i'd like to keep this conversation going if you have the time.

**Comment by rusenask** [comment]
@davecheney hey, apologies. 

TLS functionality would be nice so you could mix HTTP and HTTPS (passthrough) services in one ingress. 

I know that you could achieve the same thing with multiple service load balancers but having it in on ingress saves you one IP (atleast in GKE) and having external load balancers in GKE costs a bit. 

At least in nginx-ingress it have this and I thought it's quite nice as you can have HTTP+HTTPS+GRPC services defined in one ingress. The only downside is that nginx keeps killing long connections. I don't need contour to provision TLS certs as I request them from my backend.

**Comment by davecheney** [comment]
Thanks for your reply. I'll be honest, I'm not sure I understand what you're asking for, but I do understand supporting the
```
ingress.kubernetes.io/ssl-passthrough: "true"
```
annotation. So I'll work on that and we can see if there is anything left to do afterwards.

**Comment by davecheney** [comment]
The way I think this can be accomplished is to use a [tcp proxy][0] filter on the tls listener.

There are a number of caveats with this approach:

- It relies on TLS/SNI for host name routing. You can't do this on normal port 80 without binding to a new port.
- Envoy _must_ do TLS negotiation before making the backend connection. I don't know how to forward the TLS negotiation to the cluster backend, that seems to defeat the SSL spec. The backend service will have a TLS connection, but that TLS connection will be between envoy and itself, not the remote client.

If someone wants to attempt this, I'll mark it for 0.5, but I think the limitations are severe enough that you'd be better off using a service loadbalancer in TCP mode, you're not getting much value by adding Envoy into the mix.

[0]: https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/filter/network/tcp_proxy/v2/tcp_proxy.proto.html#envoy-api-msg-config-filter-network-tcp-proxy-v2-tcpproxy

**Comment by davecheney** [comment]
I'm going to remove the milestone from this as i'm not sure how to implement it within the framework that the k8s Ingress document provides.

**Comment by roadrunner** [comment]
this is only supported by Haproxy with L4 TCP SNI inspection.
```
frontend fc_tls
  bind *:443
  mode tcp
  tcp-request inspect-delay 3s
  tcp-request content accept if { req_ssl_hello_type 1 }
  use_backend abc-tls if { req_ssl_sni -i abc.example.org }
  use_backend xyz-tls if { req_ssl_sni -i xyz.example.org }
  default_backend https-back
```
currently Envoy doesn't support this feature but there is an old issue still active and @PiotrSikora working on a [PR](https://github.com/envoyproxy/envoy/issues/1843#issuecomment-371352012)..

**Comment by davecheney** [comment]
Blocked on envoyproxy/envoy#1843

**Comment by glerchundi** [comment]
Now that envoyproxy/envoy#1843 is already addressed and released on Envoy 1.7.0, it can be done once #443 gets fixed. Before doing anything and knowing that Contour is taking the use `IngressRoute CRD` approach (thus not polluting standard ingress with a myriad of annotations), what about something like this? 

```
spec:
  virtualhost:
    fqdn: "www.google.com"
    tls:
      passthrough: true
```

I don't know if it's even possible but appears to be easily understandable.

**Comment by glerchundi** [comment]
@davecheney, @rosskukulinski can we push this forward knowing that `IngressRoute` (project) is already more or less well-established?

**Comment by davecheney** [comment]
Thanks for the ping. I recently finished the upgrade of Envoy to 1.7.0 for the 0.7 milestone so I think we have all the raw materials needed to do this.

However, 
```
spec:
  virtualhost:
    fqdn: "www.google.com"
    tls:
      passthrough: true
```
This isn't the only config that would change. K8s services, what envoy calls clusters, are tired implicitly to routes in the ingressroute spec. Some thought about how to change the routes... section of the spect will probably be needed.

**Comment by glerchundi** [comment]
As it seems to be declared as out-of-scope in #787, when do you think would be the best release/time window to implement this? I'm open to collaborate on this.

**Comment by davecheney** [comment]
I know you've been waiting a long time for this so I want to try to give you the most complete explanation of the current status.

#787 will add TCP port forwarding for TLS connections. They will be decrypted at the Envoy edge and the SNI handshake used to direct the unencrypted traffic to the target pods.

This ticket is a request to add nginx's `ssl-passthrough` option. Having read the docs for this feature, https://kubernetes.github.io/ingress-nginx/user-guide/tls/#ssl-passthrough, I _believe_ that #787 will add the support to Contour to enable this.

However, envoyproxy/envoy#1843, suggests that it is possible to do the SNI handshake, but somehow forward the encrypted traffic to the backend service. From my understanding of the nginx documentation this is _not_ now `ssl-passthrough` works. But for completeness the nginx docs don't specify if the traffic inside the cluster is decrypted or not.

Given the ambiguity, I forwarding encrypted traffic through Envoy, what I believe this ticket is asking for, is not in scope for Contour 0.8, but obviously doing most of this plumbing will make it possible to add it in the future.

**Comment by PiotrSikora** [comment]
@davecheney `ssl-passthrough` is implemented by peeking at the TLS `ClientHello` message without consuming it (i.e. without performing TLS handshake at the proxy), and simply forwarding encrypted TCP packets between TLS endpoints, in both: Envoy (using [TLS Inspector](https://www.envoyproxy.io/docs/envoy/latest/configuration/listener_filters/tls_inspector)) and NGINX (using [`ngx_stream_ssl_preread_module`](http://nginx.org/en/docs/stream/ngx_stream_ssl_preread_module.html)).

The only difference between this and #787 (assuming you're using filter chain matching in Envoy) is whether or not [`tls_context`](https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/listener/listener.proto.html#listener-filterchain) is included in the config.

**Comment by davecheney** [comment]
@PiotrSikora thanks for confirming.

**Comment by glerchundi** [comment]
Cannot explain better than how @PiotrSikora did, thanks! And also for the tip that the absence of `tls_context` makes envoy work as ssl passthrough.

**Comment by glerchundi** [comment]
Of course Dave, will do once I empirically verify it’s something that
works. I included the patch only to show you how it’s being done.

Good night,
